src/rules/noUnnecessaryClassRule.ts

```diff
@@ -0,0 +1,126 @@
+        type: "functionality",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_CONSTRUCTOR_ONLY = "Every member of this class is a constructor. Use functions instead.";
+    public static FAILURE_STATIC_ONLY = "Every member of this class is static. Use namespaces or plain objects instead.";
+    public static FAILURE_EMPTY_CLASS = "This class has no members.";
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new NoUnnecessaryClassWalker(sourceFile, this.ruleName, this.ruleArguments));
+    }
+}
+
+const OPTION__ALLOW_CONSTRUCTOR_ONLY = "allow-constructor-only";
+const OPTION__ALLOW_EMPTY_CLASS = "allow-empty-class";
+const OPTION__ALLOW_STATIC_ONLY = "allow-static-only";
+
+class NoUnnecessaryClassWalker extends Lint.AbstractWalker<string[]> {
+    public walk(sourceFile: ts.SourceFile) {
+        const checkIfStaticOnlyClass = (node: ts.Node): void => {
+            if (isClassDeclaration(node) && !hasExtendsClause(node)) {
+                if (node.members.length === 0 && !this.hasOption(OPTION__ALLOW_EMPTY_CLASS)) {
+                    this.addFailureAtNode(getChildOfKind(node, ts.SyntaxKind.ClassKeyword)!, Rule.FAILURE_EMPTY_CLASS);
+                    return;
+                } else if (node.members.length == 0) {
+                    return;
+                }
+
+                if (node.members.some(isConstructorWithClassDeclaration)) {
+                    return ts.forEachChild(node, checkIfStaticOnlyClass);
+                }
+
+                const allMembersAreConstructors = node.members.every(isConstructorDeclaration);
+
+                if (allMembersAreConstructors && !this.hasOption(OPTION__ALLOW_CONSTRUCTOR_ONLY)) {
+                    this.addFailureAtNode(
+                        getChildOfKind(node, ts.SyntaxKind.ClassKeyword, this.sourceFile)!, Rule.FAILURE_CONSTRUCTOR_ONLY,
+                    );
+                    return;
+                }
+
+                if (!allMembersAreConstructors && !this.hasOption(OPTION__ALLOW_STATIC_ONLY)) {
+                    for (const member of node.members) {
+                        if (isConstructorWithShorthandProps(member) ||
+                        (!isConstructorDeclaration(member) && !hasModifier(member.modifiers, ts.SyntaxKind.StaticKeyword))) {
+                            return;
+                        }
+                    }
+                    this.addFailureAtNode(getChildOfKind(node, ts.SyntaxKind.ClassKeyword, this.sourceFile)!, Rule.FAILURE_STATIC_ONLY);
+                    return;
+                }
+            }
+            return ts.forEachChild(node, checkIfStaticOnlyClass);
+        };
+        ts.forEachChild(sourceFile, checkIfStaticOnlyClass);
+    }
+
+    private hasOption(option: string): boolean {
+        return this.options.indexOf(option) !== -1;
```

Comment:
Consider parsing the options upfront and passing an object to the constructor of your Walker.